core/trino-main/src/main/java/io/trino/sql/planner/LocalExecutionPlanner.java

```diff
@@ -3109,7 +3109,15 @@ public PhysicalOperation visitRefreshMaterializedView(RefreshMaterializedViewNod
         public PhysicalOperation visitTableWriter(TableWriterNode node, LocalExecutionPlanContext context)
         {
             // Set table writer count
-            context.setDriverInstanceCount(getTaskWriterCount(session));
+            // being a single node means there is one node and one writer so
```

Comment:
> being a single node means there is one node and one writer so

Single node doesn't mean single writer (there can be multiple writers per node). 

Currently, single node partitioning is used only by system partitioning handle and it's not for insert path.
This code here only deals with local distribution, but there is also `io.trino.sql.planner.optimizations.AddLocalExchanges.Rewriter#visitTableWriter` and possibly more, see changes in https://github.com/trinodb/trino/commit/b8e4e3f035e9105838cbb3f829c4d9ce64e7a2fd
I would rather not change this code.

Could we just handle your case using dedicated constant partitioning function which would direct all rows to single writer?